r1cs/gadgets/std/src/groups/nonnative/nonnative_group_gadget.rs

```diff
@@ -0,0 +1,844 @@
+    ) -> Result<Self, SynthesisError>{
+
+        let mut to_sub = SWProjective::<P>::zero();
+
+        let mut t = base.clone();
+        let sigma = base.clone();
+        let mut result = result.clone();
+
+        let mut bit_vec = Vec::new();
+        bit_vec.extend_from_slice(bits);
+        //Simply add padding. This should be safe, since the padding bit will be part of the
+        //circuit. (It is also done elsewhere).
+        if bits.len() % 2 != 0 {
+            bit_vec.push(Boolean::constant(false))
+        }
+
+        for (i, bits) in bit_vec.chunks(2).enumerate() {
+            let ti = t.clone();
+            let two_ti = SWProjective::<P>::double(&ti);
+            let mut table = [
+                sigma,
+                sigma + &ti,
+                sigma + &two_ti,
+                sigma + &ti + &two_ti,
+            ];
+
+            //Compute constants
+            SWProjective::batch_normalization(&mut table);
+            let x_coords = [table[0].x, table[1].x, table[2].x, table[3].x];
+            let y_coords = [table[0].y, table[1].y, table[2].y, table[3].y];
+            let precomp = Boolean::and(cs.ns(|| format!("b0 AND b1_{}", i)), &bits[0], &bits[1])?;
+
+            //Lookup x and y
+            let x = NonNativeFieldGadget::two_bit_lookup_lc(cs.ns(|| format!("Lookup x_{}", i)), &precomp, &[bits[0], bits[1]],  &x_coords)?;
+            let y = NonNativeFieldGadget::two_bit_lookup_lc(cs.ns(|| format!("Lookup y_{}", i)), &precomp, &[bits[0], bits[1]],  &y_coords)?;
+
+            //Perform addition
+            let adder: Self = Self::new(x, y, Boolean::constant(false));
+            result = result.add(cs.ns(||format!("Add_{}", i)), &adder)?;
+            t = t.double().double();
+            to_sub += &sigma;
+        }
+        result = result.sub_constant(cs.ns(|| "result - sigma*n_div_2"), &to_sub)?;
+        Ok(result)
+    }
+
+    fn get_value(&self) -> Option<<Self as GroupGadget<SWProjective<P>, ConstraintF>>::Value> { 
+        unimplemented!()
+    }
+
+    fn get_variable(&self) -> Self::Variable {
+        unimplemented!()
+    }
+
+    fn cost_of_add() -> usize {
+        unimplemented!()
+    }
+
+    fn cost_of_double() -> usize {
+        unimplemented!()
+    }
+}
+
+// TODO: check if this implementation is correct. 
+// (Does not seem to refer to the EqGadget)
+impl<P, ConstraintF, SimulationF> PartialEq
+for GroupAffineNonNativeGadget<P, ConstraintF, SimulationF>
+    where
+        P: SWModelParameters<BaseField = SimulationF>,
+        ConstraintF: PrimeField,
+        SimulationF: PrimeField + SquareRootField
+{
+    fn eq(&self, other: &Self) -> bool {
+        self.x == other.x && self.y == other.y
+    }
+}
+
+//  TODO: a stub?
+impl<P, ConstraintF, SimulationF> Eq
+for GroupAffineNonNativeGadget<P, ConstraintF, SimulationF>
+    where
+        P: SWModelParameters<BaseField = SimulationF>,
+        ConstraintF: PrimeField,
+        SimulationF: PrimeField + SquareRootField
+{
+}
```

Comment:
check this please

Reply:
It's fine, it's a primitive trait. The empty notation is used to say that the default implementation of functions of the trait will be used